### PR TITLE
[gitlab] remove docker tag using the CI

### PR DESCRIPTION
**Pull request by ahmed-mez**
### What does this PR do?

add a gitlab step that delete a specific tag from dockerHub

### Motivation

internal reasons

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -1255,6 +1267,25 @@ latest_release:
...
     - inv -e docker.publish --signed-pull --signed-push datadog/dogstatsd:${RELEASE} datadog/dogstatsd:latest
 
+#
+# Use this step to delete a tag of a given image
+# - remove the leading dot from the name
+# - set the IMAGE and TAG envvars
+# - in the gitlab pipeline view, trigger the step (in the first column)
+#
+.delete_docker_tag:
+  <<: *docker_delete_tag_job_definition
```
Comment (trishankatdatadog) [inline .gitlab-ci.yml]: Is there a reason the job definition should be separated like this? I don't see it being used elsewhere, so might make readability easier to just include it here? 🙂

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -1255,6 +1267,25 @@ latest_release:
     - inv -e docker.publish --signed-pull --signed-push datadog/agent:${RELEASE}-jmx datadog/agent:latest-jmx
     - inv -e docker.publish --signed-pull --signed-push datadog/dogstatsd:${RELEASE} datadog/dogstatsd:latest
 
+#
+# Use this step to delete a tag of a given image
+# - remove the leading dot from the name
```
Comment (trishankatdatadog) [inline .gitlab-ci.yml]: Uncommenting and commenting a job name like this seems fragile and error-prone. I understand we don't want to accidentally delete images, but is there a way to make this more usable?

File: .gitlab-ci.yml [hunk 3]
```diff
@@ -937,6 +937,18 @@ twistlock_scan:
...
+.docker_delete_tag_job_definition: &docker_delete_tag_job_definition
+  stage: image_deploy
+  tags: [ "runner:docker", "size:large" ]
+  image: 486234852809.dkr.ecr.us-east-1.amazonaws.com/docker-notary:0.6.1
+  before_script:
+    - DOCKER_REGISTRY_LOGIN=$(aws ssm get-parameter --region us-east-1 --name ci.datadog-agent.$DOCKER_REGISTRY_LOGIN_SSM_KEY --with-decryption --query "Parameter.Value" --out text)
+    - PASS=$(aws ssm get-parameter --region us-east-1 --name ci.datadog-agent.$DOCKER_REGISTRY_PWD_SSM_KEY --with-decryption --query "Parameter.Value" --out text)
+    - pip install -r requirements.txt
+    - |
+      export DOCKER_TOKEN=`curl -s -H "Content-Type: application/json" -X POST -d '{"username": "'$DOCKER_REGISTRY_LOGIN'", "password": "'$PASS'"}' https://hub.docker.com/v2/users/login/ | python -c 'import sys, json; print(json.load(sys.stdin)["token"].strip())'`
```
Comment (trishankatdatadog) [inline .gitlab-ci.yml]: Do we need to generate the tokens for every new job run like this? Is it because the tokens expire?

**Comment by trishankatdatadog** [comment]
It might also be helpful to future readers to document how and why this API is being called 🙂

File: .gitlab-ci.yml [hunk 4]
```diff
@@ -937,6 +937,18 @@ twistlock_scan:
     - notary -d ~/.docker/trust key import /tmp/docker.key; rm /tmp/docker.key
   dependencies: [] # Don't download Gitlab artefacts
 
+.docker_delete_tag_job_definition: &docker_delete_tag_job_definition
+  stage: image_deploy
+  tags: [ "runner:docker", "size:large" ]
+  image: 486234852809.dkr.ecr.us-east-1.amazonaws.com/docker-notary:0.6.1
```
Comment (trishankatdatadog) [inline .gitlab-ci.yml]: You don't need Notary here, amirite?

File: .gitlab-ci.yml (hunk 1 above)
Comment (ahmed-mez) [reply .gitlab-ci.yml]: yes I think in this case there is no good reason to separate the job definition, thanks for pointing out 🙂

File: .gitlab-ci.yml (hunk 2 above)
Comment (ahmed-mez) [reply .gitlab-ci.yml]: Totally agree that this is not the perfect way to enable/disable the job but I wanted to be consistant with the `latest_revert_to_previous_release` job here https://github.com/DataDog/datadog-agent/blob/d4b584b5d69220a3aa73ac2b0a83da07021fa6b2/.gitlab-ci.yml#L1251-L1268

File: .gitlab-ci.yml (hunk 3 above)
Comment (ahmed-mez) [reply .gitlab-ci.yml]: Yes, the token expires https://forums.docker.com/t/how-long-is-docker-hub-jwt-token-valid/37743

File: .gitlab-ci.yml (hunk 4 above)
Comment (ahmed-mez) [reply .gitlab-ci.yml]: Good catch, we're not signing anything so yes we don't need Notary here, do we have an alternative docker image we can use instead ? (with `aws` command installed and configured, `python` and `curl` installed)

File: .gitlab-ci.yml (hunk 2 above)
Comment (trishankatdatadog) [reply .gitlab-ci.yml]: I see. Is this well-documented? Does this work for everyone else on the team? Is there a way to not run a job unless some conditions are set?

File: .gitlab-ci.yml (hunk 3 above)
Comment (trishankatdatadog) [reply .gitlab-ci.yml]: Okay, I see, thanks! Will there be a record of the token being generated and used on the Docker Hub admin side?

File: .gitlab-ci.yml (hunk 4 above)
Comment (trishankatdatadog) [reply .gitlab-ci.yml]: IDK, but there should be :)

File: .gitlab-ci.yml (hunk 3 above)
Comment (ahmed-mez) [reply .gitlab-ci.yml]: Docker hub doesn't support that for now

File: .gitlab-ci.yml (hunk 2 above)
Comment (ahmed-mez) [reply .gitlab-ci.yml]: will work on that, thanks 👍

**Review by hkaj — COMMENTED** [review]
lgtm

**Review by hkaj — DISMISSED** [review]

**Review by trishankatdatadog — DISMISSED** [review]

File: .gitlab-ci.yml [hunk 5]
```diff
@@ -1255,6 +1255,34 @@ latest_release:
...
+#
+# Use this step to delete a tag of a given image
+# We call the Docker Hub API because docker cli doesn't support deleting tags
+# - remove the leading dot from the name
+# - set the IMAGE and TAG envvars
+# - in the gitlab pipeline view, trigger the step (in the first column)
+#
+.delete_docker_tag:
+  tags: [ "runner:docker", "size:large" ]
+  image: 486234852809.dkr.ecr.us-east-1.amazonaws.com/docker-notary:0.6.1
```
Comment (trishankatdatadog) [inline .gitlab-ci.yml]: Just one minor quibble: we shouldn't need Notary here

Comment (ahmed-mez) [reply .gitlab-ci.yml]: @hkaj do we have an alternative image with aws access and without notary?

Comment (hkaj) [reply .gitlab-ci.yml]: yeah you can use 486234852809.dkr.ecr.us-east-1.amazonaws.com/docker

Comment (ahmed-mez) [reply .gitlab-ci.yml]: thanks, I'm going to test it

Comment (ahmed-mez) [reply .gitlab-ci.yml]: @hkaj `/bin/bash: line 132: aws: command not found`

Comment (hkaj) [reply .gitlab-ci.yml]: hmm that's strange. Let's keep the notary one then if that's fine with you @trishankatdatadog and i'll have a look at what's in those images later.

Comment (ahmed-mez) [reply .gitlab-ci.yml]: I've made more tests (with notary) and added `&>/dev/null` to the `inv` command to hide its output which dismissed you reviews. Can you re-approve please?

**Review by hkaj — APPROVED** [review]

**Review by trishankatdatadog — APPROVED** [review]
